File: haskell/lists/p13.hs
-- P13 (**) Run-length encoding of a list (direct solution).

encode :: Eq x => [x] -> [Either x (Int,x)]
encode []     = []
encode (x:xs) =
  let (y,ys) = span (== x) xs in
  if ((length y)) == 0
    then (Left x) : encode ys
    else (Right ((length y) + 1, x)) : encode ys

File: ben/prolog/P01-P28-lists/p13.pl
% Run-length encoding of a list (direct solution).

count([], _, [], 0). 
count([X|Xs], X, R, N1) :-
    count(Xs, X, R, N0),
    N1 is N0 + 1.
count([X|Xs], Y, [X|Xs], 0) :- 
    X \= Y.

encode_direct([], []).
encode_direct([Y|Ys], [Y |Xs]) :-
    count([Y|Ys], Y, L, X),
    X == 1,
    encode_direct(L, Xs).
encode_direct([Y|Ys], [[X, Y] |Xs]) :-
    count([Y|Ys], Y, L, X),
    X \= 1,
    encode_direct(L, Xs).
